Rewrite ledger model privacy (#21304)
digital-asset/daml: sdk/docs/manually-written/overview/explanations/ledger-model/ledger-privacy.rst

.. wip::

   * Discuss ledger model projection with Canton projection in terms of information leakage (projection of the transaction view Merkle tree vs. as defined in here)

> The :ref:`previous section <ledger-structure>` answered what the Ledger looks like by introducing a hierarchical format to record the changes.

Reviewer: Minor, given that it's actually not the immediately previous section:
```suggestion
The :ref:`ledger structure section <ledger-structure>` answered what the Ledger looks like by introducing a hierarchical format to record the changes.
```

This section addresses the question "who sees which changes and data".
That is, it explains the privacy model for Canton Ledgers.

> The privacy model of Daml Ledgers is based on a **need-to-know basis**,

Reviewer: One row above they were Canton Ledgers. Can we use the same term?

and provides privacy **on the level of subtransactions**.
Namely, a party learns only those parts of ledger changes that affect contracts in which the party has a stake,
and the consequences of those changes.
The hierarchical structure is key here because it yields a natural notion of sub-transaction privacy.
To make this precise, the concepts of informee and witness are needed.

Informees
*********

> A party take different roles in Daml templates and choices:

Reviewer: ```suggestion
A party can take different roles in Daml templates and choices:
```

> they can be declared as contract or choice ``observer``, ``signatory`` or ``controller``.

Reviewer: I find "they" misleading, because: 1. party is singular; 2. there are other plurals in the sentence before, making "they" ambiguous. How about "it" or "the party"?

Author: I'm using "they" in the sense of the gender-neutral singular pronoun here. But it's probably easier to just repeat "the party".

> As the name suggests, every contract and choice observer should observe changes to the contract (creation or archival) and exercises of a choice, respectively.

Reviewer: I'd convert this paragraph into bullet points, it's easier visually to find the definitions.

Author: Good idea.

A signatory is bound by a contract and thus has a stake in it;

> they should also learn when the contract is created or used.

Reviewer: Does "also" here mean they learn when the contract is created or used *in addition to learning something else*? Or "also" simply refers to "just like observers"? I assume it's the latter, and I'd either drop also (esp if we go for bullet points), or be explicit "just like observers".

> An actor of an exercise, which is the controller of the choice, has a stake in the action and should therefore see the action;

Reviewer: Earlier, observers were observing exercises of a choice. It should probably be the same for controllers, and it's probably equivalent to "seeing the action", but it's not very clear. Why not go for one term for simplicity, or be explicit that the two terms are synonyms?

Author: I'll switch to "seing the exercise". The "action" in context indeed is the exercise of a choice here.

they may not have a stake in the input contract though.

> This motivates the following definition of an **informee**, namely the set of parties that should be informed about the action.

Reviewer: I'd start a new paragraph here.

Btw, what is "this" in "this motivates"? Probably it's the differences between different roles and what they're allowed to see.

Author: this = these observations

The informees for an action are the union of the sets marked with X in the following table.

.. _def-informee:

.. list-table:: Definiton: The **informees** of an action are the union of the sets marked with X.
   :widths: 20 20 20 20 20
   :header-rows: 1

   * - Action
     - Signatories
     - Contract observers
     - Actors

>      - Choice observers

Author: That's just the terminology inherited from Daml: https://docs.daml.com/daml/stdlib/Prelude.html#class-da-internal-template-functions-haschoiceobserver-31221

>      - Choice observers

Reviewer: I know, and my question goes beyond Daml: At the terminology's inception, why do we call these choice observers and not exercise observers?

>      - Choice observers

Author: I wasn't involved in the discussion back then, so I don't know the details. However, from the perspective of Daml, "choice observer" makes sense because you declare them with the `observer` keyword on a `choice` definition.

   * - **Create**
     - X
     - X
     -
     -
   * - consuming **Exercise**
     - X
     - X
     - X
     - X
   * - non-consuming **Exercise**
     - X
     - 
     - X
     - X
   * - **Fetch**
     - X
     - 
     - X
     - 

For example, the informees of a **Create** action are the signatories and observers of the created contract,

> which are called the **stakeholders** of the contract.

Reviewer: This sentence introduces the concept of stakeholders by means of an example. I'd introduce the concept first  outside of the example, otherwise it looks like a term specific to the example.

Reviewer: That might be too complicated, but can we somehow define stakeholders before the table, and then visually show the stakeholders in the table, i.e., signatories and observers?

Author: > This sentence introduces the concept of stakeholders by means of an example. I'd introduce the concept first outside of the example, otherwise it looks like a term specific to the example.

Good idea.

> That might be too complicated, but can we somehow define stakeholders before the table, and then visually show the stakeholders in the table, i.e., signatories and observers?

I'd rather keep the the two columns "signatories" and "contract observers" rather than replacing "contract observers" with "stakeholders" because otherwise it looks odd to me that stakeholders already include the signatories (even though signatories and contract observers don't have to be disjoint either, but that's a technicality I'm happy to gloss over here).

For consuming **Exercise** actions, the informees consist of the stakeholders of the consumed contract, the action's actors and choice observers.

As a design decision, a contract observer of the input contract is not informed about non-consuming **Exercise** and **Fetch** actions,
unless they are explicitly among the actors or choice observers.
This is because such actions do not change the state of the contract itself.

.. note::

>    Templates can declare :externalref:`pre-consuming <preconsuming-choices>` and :externalref:`post-consuming <postconsuming-choices>` choices.

Reviewer: Question: If a pre-consuming choice first archives the contract, can it do anything else?

   Daml compiles such choices to a non-consuming choice whose first or last consequence exercises the ``Archive`` choice on the template.
   Accordingly, contract observers are only informees of the ``Archive`` subaction, but not of the main ``Exercise`` action itself.

> In the :ref:`running example <da-ledgers-running-example>`, the ``AcceptAndSettle`` action and :ref:`its subactions <da-ledger-subaction>` have the informees shown in the blue hexagons of the next figure.

Reviewer: Very minor: I'd introduce this a bit differently, e.g., "To illustrate the concept of informees, we use the running example in the ledger structure section.[..]". Now that I'm writing it, how about we give some name to this example, just like we used to have the painter example? Maybe if the assets in the example had a more distinctive name than currencies, then we could call it "the x DvP example".

Author: > Also, you probably wanted to link to the "running workflow example", because that's where we have actual parties.

Yes. This makes sense.

> Very minor: I'd introduce this a bit differently, e.g., "To illustrate the concept of informees, we use the running example in the ledger structure section.[..]". 

Done.

> Now that I'm writing it, how about we give some name to this example, just like we used to have the painter example? Maybe if the assets in the example had a more distinctive name than currencies, then we could call it "the x DvP example".

I'll use `Alice and Bob swapping their assets` for now.

   Accordingly, contract observers are only informees of the ``Archive`` subaction, but not of the main ``Exercise`` action itself.

To illustrate the concept of informees, we use the :ref:`running example of Alice and Bob swapping their assets <ledger-structure_running_example>`.
The ``AcceptAndSettle`` action and :ref:`its subactions <da-ledger-subaction>` have the informees shown in the blue hexagons of the next figure.
For example, Alice is an informee of the root action ① because she is a signatory of the input contract #3, and Bob is an informee because he is the actor of the choice.
Similarly, Bank 2 and Bob are informees of the **Fetch** action ③ because Bank 2 is a signatory of the input contract #2 and Bob is the actor of the action.
Had Bob not been the actor, he would not be an informee because contract observers are not automatically informees of non-consuming exercises and fetches.

.. https://lucid.app/lucidchart/3176adad-0474-4755-bfb5-e323e1a65fab/edit
.. image:: ./images/dvp-acceptandsettle-informees.svg
   :align: center
   :width: 100%
   :alt: The informees of the ``AcceptAndSettle`` action and its subactions.

.. _da-model-projections:

> Projections

Reviewer: I'd have two subheadings within this section: Projections for transactions and Projections for ledgers.

***********

Informees should see the changes they are interested in,
but this does not mean that they have to see the entirety of any transaction that includes such a change.
This is made precise through *projections* of a transaction,
which define the view that a group of parties gets on a transaction.
Intuitively, given a transaction within a commit, a group of parties will see only the subtransaction consisting of all actions on contracts

> whose informees overlap with the parties. Thus, privacy is obtained on the subtransaction level.

Author: This sentence doesn't parse for me unfortunately. By definition, an action includes its consequences. So the root action of a subtransaction tree is the same as the subaction. Let's discuss this; there seems to be a more fundamental misunderstanding here.

the projection to either Alice or Bob or both consists of the whole Exercise action.
As an Exercise action contains the consequences,
Alice and Bob each see all the subactions, even if they are not an informee of the subaction itself.
For example, Alice's projection includes the Fetch subaction, Bob's ``Transfer`` exercise of on #2, and the creation of Bob's ``SimpleAsset`` contract #5.
Similarly, Bob's projection includes Alice's ``Transfer`` Exercise on #1 and the creation of Alice's ``SimpleAsset`` contract #6.

In contrast, the banks are *not* informees of the root action.
In fact, Bank 1 appears as an informee only in the ``Transfer`` Exercise action on #1 and its subaction, the creation of Bob's new asset #5.
Accordingly, the projection to Bank 1 consists of just this Exercise action.
Bank 2 appears as an informee of two unrelated actions in the tree: the Fetch action and the ``Transfer`` Exercise action on #2.
The projection to Bank 2 therefore consists of a transaction with these two actions as root actions.
This shows that projection can turn a single root action into a list of subactions.

Note the privacy implications of the banks' projections.

> While the banks learns that a ``Transfer`` has occurred from Alice to Bob or vice versa,

Reviewer: ```suggestion
While each bank learns that a ``Transfer`` has occurred from Alice to Bob or vice versa,
```

each bank does *not* learn anything about *why* the transfer occurred.
In particular, Bank 2 does not learn what happens between the Fetch and the Exercise on contract #2.
In practice, this means that Bank 1 and Bank 2 do not learn what Alice and Bob is exchanging their asset for,
providing privacy to Alice and Bob with respect to the banks.

.. _def-tx-projection:

> Formally, the **projection** of a transaction `tx = act`\ :sub:`1`\ `, …, act`\ :sub:`n` for a set `P` of parties is the

Author: > But that's not quite accurate, is it? P sees act and its subtree.

Again, `act` contains its consequences. So `act` is the whole subtree.

Do you have a suggestion how to make this easier to grasp?

subtransaction obtained by doing the following for each action `act`\ :sub:`i`:

> #. If `P` overlaps with the informees of `act`\ :sub:`i`, keep `act`\ :sub:`i` as-is.

Reviewer: Does "overlap" here mean "include"?

Author: No. Two sets overlap if they have at least one element in common. Since this concept seems confusing, I'll try to avoid "overlap" and spell it out pointwise.

#. Else, if `act`\ :sub:`i` has consequences, replace `act`\ :sub:`i` by the projection (for `P`) of its consequences,
   which might be empty.
#. Else, drop `act`\ :sub:`i`.

.. _da-model-ledger-projection:

Finally, the **projection of a ledger** `l` for a set `P` of parties is a DAG of transactions obtained as follows:

* Project the transaction of each commit in `l` for `P`.

* Remove empty transactions from the result.

* Add an edge between two (non-empty projected) transactions `tx`:sub:`1` and `tx`:sub:`2`
  if `tx2` uses a contract created by or used in `tx`:sub:`1` whose stakeholders overlap with `P`.

Notably, the projection of a ledger is not a ledger, but a DAG of transactions.
Its edges are a subset of the edges between the commits in the original ledger.

> The subtleties of this subset construction are discussed in the :ref:`causality section <local-ledger>`.

Author: Yes. I haven't updated the causality section yet (that's why it has the WIP warning at the top). When I get around to that in the next couple of (weeks/months/years), this statement here will be correct eventually ;-)

   as she is not a stakeholder of Bob's ``SimpleAsset`` of 1 USD.
   This transaction is not present in Alice's projection at all.
   Yet, the output of this transaction (contract #2) is used
   in the last commit of Alice's projection.
   Accordingly, contract #2 is shown as an input to the left, outside of the ledger.
   This effect is discussed below under :ref:`input divulgence <da-model-divulgence>`.

#. Bob's projection is analogous to Alice's:
   He sees everything of the second, third, and forth commit,
   but nothing of the first commit and instead merely contract #1 as an input.

#. Banks 1 and 2 only see the commits in which they create their ``SimpleAsset`` and the ``Transfer`` Exercises on them.
   Additionally, Bank 2 sees the Fetch of the ``SimpleAsset`` in the last commit, as already discussed above for transaction projections.

Witnesses
*********

The projection of a transaction or ledger for a set of parties `P` includes subactions whose informees are disjoint from `P`.
For example, Alice sees the Fetch action on Bob's ``SimpleAsset`` (contract #2)
because it is a consequence of Bob's ``Accept`` Exercise and Alice is an informee of this Exercise.

> Such parties are called witnesses.

Reviewer: Nit
```suggestion
Parties such as Alice are called witnesses.
```

Formally, for a given transaction `tx`, the **witnesses** of an subaction `act` of `tx` are all the parties
whose projection of `tx` contains `act` as a subaction.
Or equivalently, the witnesses of `act` are the union of the informees of all subactions of `tx` that contain `act`.
In particular, every informee of `act` is also a witness.

In terms of privacy, the witnesses are all those parties who learn about a particular action.

> Notably, the informees of an action may not learn about all its witnesses.

Reviewer: ```suggestion
Notably, the informees of an action may not learn about all witnesses of an action.
```

For example, Bank 1 is an informee of the Fetch action and Alice is a witness,
but Bank 1 cannot learn this from its projection.
This is crucial from a privacy perspective
as it hides who is involved in the hidden parts of the transaction.
In particular, this also explains why the projection of a commit is a transaction:

> the requesters cannot be retained.

Reviewer: I don't understand this last point. Could you rephrase or give an example?

.. note::
   Alice is not a witness of the Create action for Bob's ``SimpleAsset`` (contract #2)

>    although the contract appears as an input to Alice's projection.

Reviewer: This was a bit of a surprise to me in this section, because:
- It contradicts the statement above "In terms of privacy, the witnesses are all those parties who learn about a particular action"
- It concludes a section about witnesses with a statement about non-witnesses
- It seems to want to say: "There are non-witnesses that learn about an action through other mechanisms." That's fine in itself as a statement to link to the next section, especially if it's rephrased to not contradict "In terms of privacy [..]". But I wouldn't give the example here. I'd give it in the divulgence section below.

Author: Let's discuss this offline.

* **Immediate divulgence** refers to witnesses seeing contract creations they are not an informee of.
  In the example of :ref:`ledger projections of the DvP <da-dvp-ledger-projections>`,
  Bob is a witness of the Create action for Alice's new ``SimpleAsset`` (contract #6), but not an informee.
  Conceptually, at the instant where Bob exercises the ``Transfer`` choice,
  he also gains a temporary stake in the outcome of the ``Transfer``,
  namely to see that the asset now belongs to Alice.
  
  In general, there is no point in hiding the consequences of an action.
  Bob could anyway compute the consequences of the actions it is an informee of, because Daml is deterministic.

* **Input divulgence** refers to an input contract being shown to the non-informee witnesses of an action using this contract.
  For example, the ``Fetch`` on Bob's ``SimpleAsset`` (contract #2) is visible to Alice
  and Alice's projection therefore references this contract as an input
  even though the Create action for #2 is not part of Alice's projection.

  Input divulgence enables Alice to validate the transactions in her projection

>   (see :ref:`da-model-consistency` for ledger integrity).

Reviewer: Sam as above: the link uses old examples.

  That is, Alice can check that Bob does allocate a suitable ``SimpleAsset`` according to what she specified in her proposal.  

Immediate divulgence is accessible on the Ledger API via the update tree stream.

> In contrast, contracts exposed via input divulgence cannot be accessed on the Ledger API.

Reviewer: The passive doesn't help here: Could you specify to whom immediate divulgence is accessible (and what does accessible mean?). Also, who cannot access via LAPI the contracts exposed via input divulgence?

Author: I've rephrased it. Please take a look.

Disclosure is about such parties using contracts in their own transactions.

Recall from the :ref:`running example <ledger-structure_running_example>`
that Bob uses ``submitWithDisclosures`` for the exercising ``Settle`` choice.
This is because Bob (and its Participant Node) in general does not know about the ``SimpleAsset`` contract #2
that Alice has allocated to the proposal.
Disclosure means that Alice tells Bob via an off-ledger communication channel about this contract.
In the Daml script running example, the script itself is the communication channel.
In real-world contexts, Alice would offer an API for Bob to retrieve the relevant data.

It is a design decision that immediate divulgence does not entail disclosure.
For example, after the DvP has been settled, Alice creates another DvP proposal for Bob to swap the two assets again:

.. literalinclude:: ./daml/SimpleDvP.daml
   :language: daml
   :start-after: SNIPPET-REVERT-PROPOSAL-BEGIN
   :end-before: SNIPPET-REVERT-PROPOSAL-END

Then, Bob must still disclose Alice's ``SimpleAsset`` even though Bob has witnessed the creation of Alice's ``SimpleAsset``.

> A plain ``submit`` does not work.

Author: I've rephrased it differently. I want to point out that Bob must include the contract as a disclosure in its command, rather than emphasizing where Bob gets this information from.

In addition to the increase in actions on the Ledger,
the two approaches differ in in who learns about the parties that are informed about the contract:

* If Alice discloses her ``SimpleAsset`` to Bob via an off-ledger channel,
  only Alice and Bob need to know about this disclosure.
  So when Alice discloses the same contract to Charlie,
  Charlie does not need to know that Alice had already shown the contract to Bob,
  and Bob does not need to know that Alice is disclosing it to Charlie.

* In contrast, when Alice adds Bob as a contract observer and then subsequently adds Charlie as another observer,
  Bob as a contract observer is notified about the archival and the creation.
  Similarly, Charlie learns that Bob is an observer on the contract, too.
  That is, all stakeholders learn about each other.
  This created a privacy problem when Alice actually does not want that Bob and Charlie know of each other.

Moveover, adding parties as observers scales poorly to large numbers,
because every observer learns about every other observer:
A Create event with `N` observers appears in the projection of at least those `N` parties,

> which is already quadratic in `N`.

Reviewer: I don't immediately see why this is quadratic.

.. wip::

   * Discuss ledger model projection with Canton projection in terms of information leakage (projection of the transaction view Merkle tree vs. as defined in here)

The :ref:`ledger structure section <ledger-structure>` answered the question "What the Ledger looks like?" by introducing a hierarchical format to record the changes.

> This section addresses the question "who sees which changes and data".

Reviewer: V. minor, more for consistency with the previous question.
```suggestion
This section addresses the question "Who sees which changes and data?".
```

That is, it explains the privacy model for Canton Ledgers.

The privacy model of Canton Ledgers is based on a **need-to-know basis**,
and provides privacy **on the level of subtransactions**.
Namely, a party learns only those parts of ledger changes that affect contracts in which the party has a stake,
and the consequences of those changes.
The hierarchical structure is key here because it yields a natural notion of sub-transaction privacy.
To make the sub-transaction privacy notion precise, we introduce the concepts of *informee* and *witness*.

Informee
********

A party can take different roles in Daml templates and choices;

> the party can be declared as contract or choice ``observer``, ``signatory`` or ``controller``.

Reviewer: Avoids confusion:
```suggestion
the party can be declared as contract ``signatory``, ``controller``, or contract or choice ``observer``.
```

To illustrate the concept of informees, we use the :ref:`running example of Alice and Bob swapping their assets <ledger-structure_running_example>`.
The nodes in the ``AcceptAndSettle`` action have the informees shown in the blue hexagons of the next figure.
For example, Alice is an informee of the root node ① because she is a signatory of the input contract #3, and Bob is an informee because he is the actor of the choice.
Similarly, Bank 2 and Bob are informees of the Fetch node ③ because Bank 2 is a signatory of the input contract #2 and Bob is the actor.
Had Bob not been the actor, he would not be an informee because contract observers are not automatically informees of non-consuming Exercises and Fetches.

.. https://lucid.app/lucidchart/3176adad-0474-4755-bfb5-e323e1a65fab/edit
.. image:: ./images/dvp-acceptandsettle-informees.svg
   :align: center
   :width: 100%
   :alt: The informees of the nodes in the ``AcceptAndSettle`` action.

Importantly, nodes cannot exist without their children on the Ledger, as mentioned in the :ref:`ledger structure section <actions-hierarchical-structure>`;

> only actions can.

Author: I'm happy to recall that actions are subtrees, but there are no actions inside a node.

An informee is therefore informed not only about a node,
but about the whole action corresponding to this node, that is, the whole subtree rooted at this node.

> Accordingly, the informees for an action are the informees of its root node.

Author: They are not informees of the nodes, neither of the actions rooted at these nodes. They are merely witnesses to these nodes. I'll rewrite this paragraph as a whole as it clearly misled you.

Here, the Create node ③ is part of three subactions, namely those rooted at nodes ①, ②, and ③.
Accordingly, this Create node is shown to the informees of all these actions, even if they are not informees of the node itself.
Those parties are called witnesses.
Formally, for a given transaction `tx`, the **witnesses** of a node in `tx` are the union of the informees of all subactions of `tx` that contain the node.
In particular, every informee of the node is also a witness.

The diagram shows the witnesses of a subaction in purple on its root action.
For node ③, the witnesses are Alice, Bob, and Bank 1, because Bob is an informee of ① and ③; Bank 1 is an informee of ② and ③; and Alice is an informee of ① and ②.

.. https://lucid.app/lucidchart/e5ee4a5f-3dda-4eec-a773-0ab3dde2c4ff/edit
.. image:: ./images/dvp-settle-witness.svg
   :align: center
   :width: 60%
   :alt: The informees of the subactions of the ``Settle`` choice and the witnesses of the nodes.

.. _da-model-projections:

Projection
**********

Informees should see the changes they are interested in,

> but this does not mean that they have to see the entirety of any transaction that includes such a change.

Reviewer: ```suggestion
but this does not mean that they are entitled to see the entirety of any transaction that includes such a change.
```

This is made precise through *projections* of a transaction,
which define the view that a group of parties gets on a transaction.
Intuitively, given a transaction within a commit, a group of parties sees only the subtransaction consisting of all actions on contracts
whose informees include at least one of the parties.
Equivalently, the group of parties sees only those nodes whose witnesses include at least one of the parties.
Thus, privacy is obtained on the subtransaction level.

This section first defines projections for transactions and then for ledgers.

Transaction projection
======================

The next diagram gives an example for a transaction with the ``AcceptAndSettle`` Exercise action as the only root action, whose informees are shown above.

.. https://lucid.app/lucidchart/9b3762db-66b4-4e72-94cb-bcefd4c1a5ea/edit

> .. image:: ./images/dvp-acceptandsettle-projection.svg

Reviewer: V minor and probably not worth the effort: the number on the nodes would be nice to have.

Author: I'd rather keep them out of the projections because I don't see how I can get them right:

* Option 1: Keep the numbers consistent across projections.
* Option 2: Renumber the nodes in each projection.

If I go for option 1, then a reader might misinterpret these numbers as being part of the projection. So we could use them to figure out the order of interleaving in the projections for Bank 1/2 from the individual projections. That's wrong. Moreover, a reader might confuse them with the event IDs available on the ledger API and assume that those IDs are stable across projections, which is wrong.

If I go for option 2, then I fear that this will just be more confusing because node 6 from the original transaction will be node 1 in Bank 1's projection and node 2 in the joint projection for Bank 1/2.

namely Bob exercising the ``AcceptAndSettle`` choice on Alice's ``ProposeSimpleDvP`` contract,
the projection to either Alice or Bob or both consists of the whole Exercise action.
As an action consists of the whole subtree underneath its root node,
Alice and Bob each see all the nodes they are witnesses of.
For example, Alice's projection includes the Fetch subaction, Bob's ``Transfer`` exercise of on #2, and the creation of Bob's ``SimpleAsset`` contract #5.
Similarly, Bob's projection includes Alice's ``Transfer`` Exercise on #1 and the creation of Alice's ``SimpleAsset`` contract #6.

In contrast, the banks are *not* informees of the root action.
In fact, Bank 1 appears as an informee only in the ``Transfer`` Exercise action on #1 and its subaction, the creation of Bob's new asset #5.
Accordingly, the projection to Bank 1 consists of just this Exercise action.
Bank 2 appears as an informee of two unrelated actions in the tree: the Fetch action and the ``Transfer`` Exercise action on #2.
The projection to Bank 2 therefore consists of a transaction with these two actions as root actions.
This shows that projection can turn a single root action into a list of subactions.

> Note the privacy implications of the banks' projections.

Reviewer: V minor: I'd make this paragraph an actual `.. note:`, because it's quite central.

While each bank learns that a ``Transfer`` has occurred from Alice to Bob or vice versa,
each bank does *not* learn anything about *why* the transfer occurred.
In particular, Bank 2 does not learn what happens between the Fetch and the Exercise on contract #2.
In practice, this means that Bank 1 and Bank 2 do not learn what Alice and Bob is exchanging their asset for,

> providing privacy to Alice and Bob with respect to the banks.

Reviewer: V minor: I'd make this paragraph an actual .. note:, because it's quite central.

Author: Good idea.

> The projection to Bank 1 and Bank 2 at the bottom shows that a projection to several parties may contain strictly more information than the projections to each of the parties together.

Reviewer: The projection for Bank2 is the same as the projection for both, but if this is what you meant, then perhaps be explicit about it?

> Said differently, it is possible to reconstruct the projection to Bank 1 and Bank 2 from solely from the projection for Bank 1 and the projection for Bank 2.

Reviewer: ```suggestion
Said differently, it is possible to reconstruct the projection to Bank 1 and Bank 2 solely from the projection for Bank 1 and the projection for Bank 2.
```

> Said differently, it is possible to reconstruct the projection to Bank 1 and Bank 2 from solely from the projection for Bank 1 and the projection for Bank 2.

Reviewer: Didn't you mean to say that it is **not** possible, because we lose ordering information in the individual projections compared to a projection for a set of parties?
```suggestion
Said differently, it is not possible to reconstruct the projection to Bank 1 and Bank 2 from solely from the projection for Bank 1 and the projection for Bank 2.
```

Author: Indeed. Thanks for catching this.

> Here, this is because the order of the three root actions cannot be uniquely determined from the individual projections.

Author: Bank 1's projection for the transaction contains one root node (Exercise on `#1`).
Bank 2's projection for the transaction contains two root nodes (Fetch on `#2` and Exerdcise on `#2`). 
This makes three root nodes in total that appear in their joint projection.

For this reason, projection is defined for a set of parties.

.. _def-tx-projection:

Formally, the **projection** of a transaction for a set `P` of parties is the

> subtransaction obtained by doing the following for each root action `act` of the transaciton.

Reviewer: Could we make this a definition using a keyword?

Author: Good idea.

* Add an edge between two (non-empty projected) transactions `tx`:sub:`1` and `tx`:sub:`2`
  if `tx2` uses a contract created by or used in `tx`:sub:`1` whose stakeholders overlap with `P`.

The edges of the DAG are a subset of the edges between the commits in the original ledger.
The subtleties of this subset construction are discussed in the :ref:`causality section <local-ledger>`.
Until then, we pretend that the ledger is totally ordered and projections retain the same ordering.

Notably, the projection of a ledger is not a ledger, but a DAG of transactions.

> The requesters cannot be retained because they are typically witnesses of the actions in the projection,

Reviewer: What are the requesters?

Author: The set of parties mentioned in a commit. See the ledger-structure section.

  In general, there is no point in hiding the consequences of an action.
  Bob could anyway compute the consequences of the actions it is an informee of, because Daml is deterministic.

* **Input divulgence** refers to an input contract being shown to the non-informee witnesses of a node using this contract.
  For example, the Fetch on Bob's ``SimpleAsset`` (contract #2) is visible to Alice
  and Alice's projection therefore references this contract as an input
  even though the Create action for #2 is not part of Alice's projection.

  Input divulgence enables Alice to validate the transactions in her projection
  (see :ref:`da-model-consistency` for ledger integrity).
  That is, Alice can check that Bob does allocate a suitable ``SimpleAsset`` according to what she specified in her proposal.

  Input divulgence does not make Alice a witness of the Create action for Bob's ``SimpleAsset`` (contract #2),

>   because an input contract is not the same as its Create action.

Reviewer: In what sense is an input contract different from its Create action (apart from how it's depicted in the diagrams)? I don't know if it's relevant to expand in the text, but I'm curious.

Author: I don't want to go into the technical details in this doc, but for your reference:

An input contract is technially a `FatContractInstance` https://github.com/digital-asset/daml/blob/ace3d6af682417257bbaa475dc4d31bf87406cb8/sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/FatContractInstance.scala#L14.

A create action consists only of a create node https://github.com/digital-asset/daml/blob/ace3d6af682417257bbaa475dc4d31bf87406cb8/sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala#L82.

There are methods to convert between the two, but they are not lossless:

https://github.com/digital-asset/daml/blob/ace3d6af682417257bbaa475dc4d31bf87406cb8/sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/FatContractInstance.scala#L34-L43 

https://github.com/digital-asset/daml/blob/ace3d6af682417257bbaa475dc4d31bf87406cb8/sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/FatContractInstance.scala#L105-L122

Moreover, a Create node typically exists inside a tree (and as such has an event ID relative to the update ID) which doesn't make sense for an input contract. Unfortunately, the LAPI blurs this distinction by including the update/event ID of the last change to a contract in the ACS snapshot.

  In the diagrams, this distinction is visualized via the dashed border for input contracts and them being placed to the left.
  
Via the Ledger API's :externalref:`update service <com.daml.ledger.api.v2.GetUpdateTrees>`,
a user can see the immediately divulged contracts in the trees of the parties' projection

> as these trees contain the Create nodes.

Reviewer: I'm not sure what you mean: Can a user see that in Alice's projection contract #2 is divulged because the Alice's tree contains the Create of contract #2? I thought Alice's tree doesn't contain it, and that was in fact the different between divulging as input and seeing the Create,

Author: Contract `#2` is input-divulged to Alice, not immediately divulged. Are you talking about contract `#5`?

> In contrast, the Ledger API currently does not offer a means for a user to lookup a contract ID of an input divulgence.

Reviewer: ```suggestion
In contrast, the Ledger API currently does not offer a means for a user to look up a contract ID of an input divulgence.
```

In the Daml script running example, the script itself is the communication channel.
In real-world contexts, Alice would offer an API for Bob to retrieve the relevant data.

It is a design decision that immediate divulgence does not entail disclosure for subsequent transactions.
For example, after the DvP has been settled, Alice creates another DvP proposal for Bob to swap the two assets again:

.. literalinclude:: ./daml/SimpleDvP.daml
   :language: daml
   :start-after: SNIPPET-REVERT-PROPOSAL-BEGIN
   :end-before: SNIPPET-REVERT-PROPOSAL-END

> Then, Bob's command submission must include the disclosure of Alice's ``SimpleAsset`` even though Bob has witnessed the creation of Alice's ``SimpleAsset``.

Reviewer: This makes it clearer that we use "witness" in the sense defined in this section.

```suggestion
Then, Bob's command submission must include the disclosure of Alice's ``SimpleAsset`` even though Bob is a witness of the creation of Alice's ``SimpleAsset``.
```

A plain ``submit`` without disclosure does not work.

> The motivation for using immediate divulgence implicitly for disclosure is that it leads to brittle workflows.

Reviewer: Shouldn't it be "for **not**"?
```suggestion
The motivation for not using immediate divulgence implicitly for disclosure is that it leads to brittle workflows.
```
